mbotmake2/extract_thumbnails.py

```diff
@@ -8,23 +8,68 @@
 
 def extractThumbnails(filename: Path, output_dir: Path) -> list[Path]:
     with open(filename, "r") as file:
+        # Read the file in chunks to extract only relevant parts containing THUMBNAIL_BLOCK_START or THUMBNAIL_BLOCK_END
+        # Load first 5MB of the file
+        file.seek(0)
+        content = file.read(5_000_000)
+
+        # It seems that these are pretty standard for thumbnail blocks
+        # Orcaslicer uses '; THUMBNAIL_BLOCK_START' and '; THUMBNAIL_BLOCK_END' additionally to wrap the thumbnail block.
+        THUMBNAIL_BLOCK_START = "; thumbnail begin"
+        THUMBNAIL_BLOCK_END   = "; thumbnail end"
+
+        currentlyCollectingThumbnail = False
+        countCollectedThumbnail = 0
+        relevant_parts = []
+        for line in content.splitlines():
+            if THUMBNAIL_BLOCK_START in line:
+                currentlyCollectingThumbnail = True
+                #print(f"Found '{THUMBNAIL_BLOCK_START}'. Starting to collect thumbnail data.")
+
+            if currentlyCollectingThumbnail:
+                if line.strip() and line.strip() != ";":
+                    relevant_parts.append(line)
+
+            if THUMBNAIL_BLOCK_END in line:
+                currentlyCollectingThumbnail = False
+                countCollectedThumbnail += 1
+                #print(f"Found '{THUMBNAIL_BLOCK_END}'. Stopped collecting thumbnail data.")
+
+        # Combine the relevant parts for parsing
+        relevant_content = "\n".join(relevant_parts) + "\n"
+
+        #print(f"Extracted {len(relevant_parts)} lines of thumbnail data.")
+        #print("Relevant content for parsing:")
+        #print(relevant_content)
+
         try:
-            ast = grammar.match(file.read(5_000_000))
-        except ParseError:
-            print("Thumbnail not found. Skipping thumbnail generation...")
+            print("Attempting to parse the extracted thumbnail blocks...")
+            ast = grammar.match(relevant_content)
+            #print("Parsing successful. AST generated.")
+        except ParseError as e:
+            print("No thumbnails were not found or parsing failed. Skipping thumbnail generation...")
+            print(f"ParseError: {e}")
```

Comment:
Again, `relevant_content` seems to defeat the PEG grammar definition and parsing mechanism. Whenever we write our own parsing logic nested on top of the PEG, it makes our unit testing at `tests/test_thumbnail.py` much harder to flush out the bugs. 

I can show you how the code `relevant_content` is done in PEG. Please refer to pull request #18 .